libp2p/pubsub/gossipsub.py

```diff
@@ -528,82 +528,105 @@ def mesh_heartbeat(
                     peers_to_prune[peer].append(topic)
         return peers_to_graft, peers_to_prune
 
+    def _handle_topic_heartbeat(
+        self,
+        topic: str,
+        current_peers: set[ID],
+        is_fanout: bool = False,
+        peers_to_gossip: DefaultDict[ID, dict[str, list[str]]] | None = None,
+    ) -> tuple[set[ID], bool]:
+        """
+        Helper method to handle heartbeat for a single topic,
+        supporting both fanout and gossip.
+
+        :param topic: The topic to handle
+        :param current_peers: Current set of peers in the topic
+        :param is_fanout: Whether this is a fanout topic (affects expiration check)
+        :param peers_to_gossip: Optional dictionary to store peers to gossip to
+        :return: Tuple of (updated_peers, should_remove_topic)
+        """
+        if self.pubsub is None:
+            raise NoPubsubAttached
+
+        # Skip if no peers have subscribed to the topic
+        if topic not in self.pubsub.peer_topics:
+            if is_fanout:
+                # For fanout topics, check if we should remove the topic
+                if self.time_since_last_publish.get(topic, 0) + self.time_to_live < int(
+                    time.time()
+                ):
+                    return set(), True
+            return current_peers, False
+
+        # For fanout topics, check if we should remove the topic
+        if is_fanout:
+            if self.time_since_last_publish.get(topic, 0) + self.time_to_live < int(
+                time.time()
+            ):
+                return set(), True
```

Comment:
It looks like the fanout check is duplicated here. Would this not achieve the same result?

```suggestion
        # For fanout topics, check if we should remove the topic
        if is_fanout:
            if self.time_since_last_publish.get(topic, 0) + self.time_to_live < int(
                time.time()
            ):
                return set(), True
                
        # Skip if no peers have subscribed to the topic
        if topic not in self.pubsub.peer_topics:
            return current_peers, False
```

Reply:
Ah yes it seems like the fanout check is duplicated. Also quick question, wouldnt it be a better choice if we first check `if topic not in self.pubsub.peer_topics` ?